File: python/Q215_kth_largest_element_in_an_array.py
#
# @lc app=leetcode id=215 lang=python3
#
# [215] Kth Largest Element in an Array
#

# @lc code=start
import heapq


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        min_heap: list[int] = []

        for n in nums:
            heapq.heappush(min_heap, n)

        while len(min_heap) > k:
            heapq.heappop(min_heap)

        return min_heap[0]


# @lc code=end

File: rust/src/leetcode/Q215_kth_largest_element_in_an_array.rs
/*
 * @lc app=leetcode id=215 lang=rust
 *
 * [215] Kth Largest Element in an Array
 */

// @lc code=start
use std::cmp::Reverse;
use std::collections::BinaryHeap;

impl Solution {
    pub fn find_kth_largest(nums: Vec<i32>, k: i32) -> i32 {
        let k: usize = k as usize;
        let mut heap: BinaryHeap<Reverse<i32>> = BinaryHeap::with_capacity(k);

        for &n in nums.iter() {
            if heap.len() < k {
                heap.push(Reverse(n));
            } else if n > heap.peek().unwrap().0 {
                heap.pop();
                heap.push(Reverse(n));
            }
        }

        heap.peek().unwrap().0
    }
}
// @lc code=end
